### PR TITLE
Add some tracing spans to give insight into local joins

### PR DESCRIPTION
**Pull request by H-Shay**
As part of the local joins investigation I'd like to add some tracing spans to get better insight into the timing of processing the `/join` request.

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -564,25 +565,26 @@ async def update_membership(
...
-                    new_room=new_room,
-                    require_consent=require_consent,
-                    outlier=outlier,
-                    historical=historical,
-                    allow_no_prev_events=allow_no_prev_events,
-                    prev_event_ids=prev_event_ids,
-                    state_event_ids=state_event_ids,
-                    depth=depth,
-                )
+                with opentracing.start_active_span("update_membership_locked"):
```
Comment (richvdh) [inline synapse/handlers/room_member.py:568]: we should probably give the Linearizer some tracing annotations at some point, which would probably make this redundant. but that's a bigger job and this is certainly fine for now.

**Review by richvdh — APPROVED** [review]
lgtm!

**Comment by H-Shay** [comment]
Thanks! Merged.

File: synapse/handlers/room_member.py (hunk 1 above)
Comment (H-Shay) [reply synapse/handlers/room_member.py:568]: Curious what that job would entail?

Comment (richvdh) [reply synapse/handlers/room_member.py:568]: really, much the same as you're doing here - adding `with opentracing.start_active_span(...)`  calls around things that might be slow. I'd probably add one at the top level of `Linearizer.queue` (or rather, in the nested `_ctx_manager`), and another at the point where we wait for the lock (`await new_defer`).

Comment (H-Shay) [reply synapse/handlers/room_member.py:568]: I think this was asked elsewhere but I didn't see the answer- is there any performance penalty or other drawback to consider when adding opentracing spans?

Comment (richvdh) [reply synapse/handlers/room_member.py:568]: Yes, there's a small performance penalty, which is one reason not to go completely crazy with adding spans. 

There are two aspects to the performance penalty. One, each call to `start_active_span` results in a few function calls and python objects. Two, if tracing is actually enabled for the current request then we additionally have to serialize all the trace data and send it over the network to the jaeger server, and obviously the more spans, the more work this is.

Plus, in some ways, less can be more when it comes to looking at a trace and understanding what's going on.

Comment (H-Shay) [reply synapse/handlers/room_member.py:568]: Makes sense, thanks for explaining!

**Pull request merged**